### PR TITLE
LCOW: Support most operations excluding remote filesystem

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

This is the start of the Linux Containers on Windows changes and implements a significant chunk of it.

This PR enables pull of a Linux image on Windows, committing a layer, basics of build, import and export. It does NOT include the remote filesystem work that is being tracked separately to allow things such as COPY or ADD to work (which need remoting to a service VM as Windows can't directly read or perform operations on a Linux filesystem).

This PR has been updated based on feedback from @crosbymichael so that when LCOW is enabled, the daemon assumes Linux-mode, and won't be able to run/see Windows containers, or see any Windows images. This update means that no API changes are required while the UX changes are being investigated separately. This is an artificial crippling with no technical reason.

park
(In memory of Guide Dogs for the Blind puppy-in-training "Park", front left. He sadly passed too young. 1st Jan 2016 - 16th June 2017. Rest in peace, sweet boy :dog: 💛

My two guys in the background, another former GDB puppy-in-training, "Palmetto" front-right )

### CODE REVIEW

**Comment by jhowardmsft** [comment]
@dmcgowan @stevvooe @crosbymichael  PTAL.

@johnstep @thaJeztah FYI.

File: daemon/graphdriver/lcow/lcow.go
Comment (crosbymichael) [inline daemon/graphdriver/lcow/lcow.go]: Can you please explain a little bit better what the dependency is that we are waiting for so that the implementation can be completed?  From the TODO, its not clear that this is still being worked on and has dependencies on continuity or other projects.

File: daemon/oci_windows.go
Comment (crosbymichael) [inline daemon/oci_windows.go]: You should have the code that fixes up a spec for LCOW factored out into a separate function and not littered though this one.  In the future its easy to know where to add LCOW spec modifications and lot have to look in multiple places.

Its only a couple options now but when you do runtime it can expand more

**Comment by jhowardmsft** [comment]
image

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Actually this has been updated in the latest push to have a (very, very minimal) Get(). There will be further work with the remote filesystem work that Akash is working with Stephen on and integrating Continuity, but that is completely separate from this PR itself.

File: daemon/oci_windows.go
Comment (jhowardmsft) [inline daemon/oci_windows.go]: Yes, this has now been re-factored into a Linux and a Windows function

**Comment by jhowardmsft** [comment]
Rebased. I'm adding more commits making further progress while waiting for review as the PR isn't making any progress after removing `WIP` for getting on for a week after discussing offline. Also moving to code review. @crosbymichael @stevvooe @dmcgowan @thaJeztah @johnstep PTAL.

**Comment by crosbymichael** [comment]
You should have pinged us in the PR after removing the WIP that you were ready.  I didn't get an email when the PR title changes.

**Comment by crosbymichael** [comment]
Restarted the failed builds

File: api/server/backend/build/tag.go
Comment (crosbymichael) [inline api/server/backend/build/tag.go]: This can be done in a separate PR but after writing these 4 lines over 10 different times, i think this can be factored out into a function for handling platform switching

**Comment by crosbymichael** [comment]
@jhowardmsft is the RS1 builders broken at the moment?

File: api/server/backend/build/tag.go
Comment (jhowardmsft) [inline api/server/backend/build/tag.go]: To be fair, all of these comments are markers on account of the fact of your request to cripple this PR such that the daemon -only- runs in either 'windows' or 'lcow' mode. If the API changes (particularly to build, run and import) are accepted to add `platform`, then pretty much all of these can be removed. The crippling is artificial, not technical. There is absolutely no need to factor them out - they are all temporary due to your request.

**Comment by jhowardmsft** [comment]
Updated to add support for commit as well.

**Comment by jhowardmsft** [comment]
And now with support for `commit` in the latest changes

image

**Comment by jhowardmsft** [comment]
@friism FYI

**Comment by crosbymichael** [comment]
The failures on janky and experimental look legitimate.  They are coming the the docker-py tests that interact with the API directly.

**Comment by jhowardmsft** [comment]
@crosbymichael Hmmm, they're not coming from docker-py, that already passed.

```
02:10:22.820 +++ exec dockerd --debug --host unix:///go/src/github.com/docker/docker/bundles/17.06.0-dev/test-docker-py/docker.sock --storage-driver vfs --pidfile bundles/17.06.0-dev/test-docker-py/docker.pid --userland-proxy=true
02:10:24.015 
02:10:24.016 ++ env -i DEST=bundles/17.06.0-dev/test-docker-py DOCKER_CLI_VERSION=unknown-version DOCKER_API_VERSION=1.30 DOCKER_INTEGRATION_DAEMON_DEST= DOCKER_TLS_VERIFY= DOCKER_CERT_PATH= DOCKER_ENGINE_GOARCH= DOCKER_GRAPHDRIVER=vfs DOCKER_USERLANDPROXY=true DOCKER_HOST=unix:///go/src/github.com/docker/docker/bundles/17.06.0-dev/test-docker-py/docker.sock DOCKER_REMAP_ROOT= DOCKER_REMOTE_DAEMON= DOCKERFILE= GOPATH=/go GOTRACEBACK=all HOME=/go/src/github.com/docker/docker/bundles/17.06.0-dev/test-docker-py/fake-HOME PATH=/go/src/github.com/docker/docker/bundles/17.06.0-dev/test-docker-py/../binary-daemon:/go/src/github.com/docker/docker/bundles/17.06.0-dev/test-docker-py/../dynbinary-daemon:/usr/local/cli:/go/bin:/usr/local/go/bin:/osxcross/target/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin TEMP= TEST_IMAGE_NAMESPACE= TEST_CLIENT_BINARY=docker PYTHONPATH=/docker-py py.test --junitxml=bundles/17.06.0-dev/test-docker-py/results.xml /docker-py/tests/integration
02:10:24.330 ============================= test session starts ==============================
02:10:24.331 platform linux2 -- Python 2.7.9, pytest-2.9.1, py-1.4.34, pluggy-0.3.1
02:10:24.332 rootdir: /docker-py, inifile: pytest.ini
02:10:24.332 plugins: cov-2.1.0
...
02:17:29.408 SKIP [1] tests/integration/api_service_test.py:300: API version is too low (< 1.27)
02:17:29.408 SKIP [1] /docker-py/tests/integration/api_swarm_test.py:24: Test stalls the engine on 1.12.0
02:17:29.409 SKIP [1] tests/integration/models_services_test.py:84: Makes Swarm unstable?
02:17:29.409 SKIP [1] tests/integration/api_image_test.py:269: Doesn't work inside a container - FIXME
02:17:29.409 ============== 242 passed, 7 skipped, 3 xfailed in 425.09 seconds ==============
```

It's either coming from make cross or make tgz - a normal janky run would look like this:

```
02:25:29.889 ---> Making bundle: cross (in bundles/17.06.0-dev/cross)
02:25:29.930 Created symlinks: bundles/17.06.0-dev/cross/linux/amd64/docker-containerd bundles/17.06.0-dev/cross/linux/amd64/docker-containerd-ctr bundles/17.06.0-dev/cross/linux/amd64/docker-containerd-ctr.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-containerd-ctr.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-containerd-shim bundles/17.06.0-dev/cross/linux/amd64/docker-containerd-shim.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-containerd-shim.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-containerd.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-containerd.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-init bundles/17.06.0-dev/cross/linux/amd64/docker-init.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-init.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-proxy bundles/17.06.0-dev/cross/linux/amd64/docker-proxy.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-proxy.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-runc bundles/17.06.0-dev/cross/linux/amd64/docker-runc.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-runc.sha256 bundles/17.06.0-dev/cross/linux/amd64/dockerd bundles/17.06.0-dev/cross/linux/amd64/dockerd-17.06.0-dev bundles/17.06.0-dev/cross/linux/amd64/dockerd-17.06.0-dev.md5 bundles/17.06.0-dev/cross/linux/amd64/dockerd-17.06.0-dev.sha256
02:28:19.086 Building: bundles/17.06.0-dev/cross/windows/amd64/dockerd-17.06.0-dev.exe
02:29:17.736 Created binary: bundles/17.06.0-dev/cross/windows/amd64/dockerd-17.06.0-dev.exe
02:29:42.702 
02:29:42.708 ---> Making bundle: tgz (in bundles/17.06.0-dev/tgz)
02:29:42.708 tgz is deprecated
02:29:42.708 
02:29:46.344 Set build name.
```

but the failure is

```
init.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-init.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-proxy bundles/17.06.0-dev/cross/linux/amd64/docker-proxy.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-proxy.sha256 bundles/17.06.0-dev/cross/linux/amd64/docker-runc bundles/17.06.0-dev/cross/linux/amd64/docker-runc.md5 bundles/17.06.0-dev/cross/linux/amd64/docker-runc.sha256 bundles/17.06.0-dev/cross/linux/amd64/dockerd bundles/17.06.0-dev/cross/linux/amd64/dockerd-17.06.0-dev bundles/17.06.0-dev/cross/linux/amd64/dockerd-17.06.0-dev.md5 bundles/17.06.0-dev/cross/linux/amd64/dockerd-17.06.0-dev.sha256
02:20:35.996 Building: bundles/17.06.0-dev/cross/windows/amd64/dockerd-17.06.0-dev.exe
02:21:36.134 Created binary: bundles/17.06.0-dev/cross/windows/amd64/dockerd-17.06.0-dev.exe
02:21:54.702 # github.com/docker/docker/vendor/github.com/Microsoft/hcsshim
02:21:54.703 vendor/github.com/Microsoft/hcsshim/hcsshim.go:151: type [1073741824]uint16 too large
02:22:04.872 Build step 'Execute shell' marked build as failure
02:22:04.872 Set build name.
02:22:04.873 New build name is '#33241'

```

What's more odd is that `vendor/github.com/Microsoft/hcsshim/hcsshim.go` hasn't changed in this PR, and hasn't changed between v0.5.17 and 0.5.23 which is the new release vendored in this PR. I'll figure it, but I'm not sure I get it at the moment.

**Comment by jhowardmsft** [comment]
Rebased after https://github.com/moby/moby/pull/33454 was merged into master

**Comment by jhowardmsft** [comment]
@crosbymichael The CI failure is actually coming from cross-compiling the `windows/386` packages. Unfortunately the CI script doesn't actually output what it's doing, so it's hidden (I'll open a PR shortly to fix the CI scripts so it's more obvious). Still tracking down though why hcsshim is being pulled in on 386.

**Comment by jhowardmsft** [comment]
After the rebase, the error actually is in docker-py now. I think I figured out the HCSShim error. Investigating.

**Comment by jhowardmsft** [comment]
Actually, the docker-py failure isn't in this PR. It seems to be affecting master all over the place. For example https://jenkins.dockerproject.org/job/Docker-PRs-experimental/35151/console for https://github.com/moby/moby/pull/33749 is failing in exactly the same way, as are others.

**Comment by jhowardmsft** [comment]
Yeah, https://github.com/moby/moby/pull/33735 is the fix for docker-py. Not related to this PR.

**Comment by jhowardmsft** [comment]
And showing the builder operating in both exec and JSON form of run

image

**Comment by jhowardmsft** [comment]
Update: https://github.com/moby/moby/pull/33757 should fix the docker-py failures currently on master.

**Comment by jhowardmsft** [comment]
Short of a rebase once 33757 is merged to alleviate the docker-py failures, this is fully reviewable in its current state. @crosbymichael @stevvooe @dmcgowan PTAL

@friism, @thaJeztah  FYI

**Comment by jhowardmsft** [comment]
Yay, green again 💚

**Comment by crosbymichael** [comment]
LGTM

File: api/server/backend/build/tag.go
Comment (johnstep) [inline api/server/backend/build/tag.go]: Even if they are removed, it would be been nice to have separate functions, something like GetPlatform (get the effective platform), IsWindowsPlatform (check if it is WCOW), and something like EnsurePlatform (make sure platform is not empty). This might seem trivial, but would allow comments and reasoning to be in one place.

Also, the request was to always be in Linux or always be in Windows mode, and I think that means platform could have been set once, globally, instead of "computed" all over the place. It was not really about crippling, but about breaking down the work into smaller phases. However, the current changes should make it easier to switch to multiple platform support, so this approach seems okay to me.

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by jhowardmsft** [comment]
@johnstep - spoke offline on slack, but yes, all these nits can easily be addressed in a followup. Thanks for reviewing!

File: api/server/backend/build/backend.go
Comment (jhowardmsft) [inline api/server/backend/build/backend.go]: No, just `string` is correct :)

File: builder/dockerfile/parser/parser.go
Comment (jhowardmsft) [inline builder/dockerfile/parser/parser.go]: Yes, fixing in a `nits` PR shortly

File: api/server/backend/build/tag.go
Comment (jhowardmsft) [inline api/server/backend/build/tag.go]: Fixed in nits PR about to be submitted

File: container/container.go
Comment (jhowardmsft) [inline container/container.go]: Fixed in nits PR about to be submitted

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Fixed in nits PR about to be submitted

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Fixed in nits PR about to be submitted

File: daemon/graphdriver/windows/windows.go
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: Fixed in nits PR about to be submitted

**Pull request merged**
